packages/playground-server/src/node.ts

```diff
@@ -40,30 +50,86 @@ export class LocalFirebaseServiceFactory extends FirebaseServiceFactory {
   }
 }
 
-export let serviceFactory: FirebaseServiceFactory;
+class StandardFirebaseServiceFactory extends FirebaseServiceFactory
+  implements IClosableFirebaseServiceFactory {
+  constructor(params: any) {
+    super(params);
+  }
+  async closeServiceConnections() {}
+}
+
+class LocalPersistentFirebaseServiceFactory extends FirebaseServiceFactory
+  implements IClosableFirebaseServiceFactory {
+  firebaseServer: FirebaseServer;
+  constructor(
+    private readonly host: string,
+    private readonly port: string,
+    snapshot: any
+  ) {
+    super({
+      databaseURL: `ws://${host}:${port}`,
+      projectId: "projectId",
+      apiKey: "",
+      authDomain: "",
+      storageBucket: "",
+      messagingSenderId: ""
+    });
+
+    this.firebaseServer = new FirebaseServer(this.port, this.host, snapshot);
+  }
+
+  static async create(host: string, port: string) {
+    const snapshot = await getPlaygroundSnapshot();
+
+    return new LocalPersistentFirebaseServiceFactory(host, port, snapshot);
+  }
+
+  async closeServiceConnections() {
+    const snapshot = await this.firebaseServer.getValue();
+    storePlaygroundSnapshot(snapshot);
+
+    await this.firebaseServer.close();
+  }
+}
+
+export let serviceFactory: Promise<IClosableFirebaseServiceFactory>;
```

Comment:
Might be better to name this something like `unresolvedServiceFactory` so that when we `await` it in its usage, we can refer to the resolution as just `serviceFactory` instead of `resolvedServiceFactory`

Small nit, not too opinionated either way but just mentioning it.

Reply:
Good point, in the end I chose to name it `serviceFactoryPromise` so it's clear that's a Promise and has to be awaited before use.